### DISCUSSION TITLE
Release 4.9.0 Tracking

### DISCUSSION THREAD

**Discussion by hazendaz** [spotbugs/spotbugs#3111] [General]
I know everyone has been busy elsewhere.  We have been overdue for a while.  Are any of the open PRs still in need to be applied to 4.9.0 or can we just cut it?  If we are in consensus to just cut this, do we need to revisit the commons-text / commons-lang issues we faced in prior release or are we good with that?  I'm sure many would like us to get this out as soon as possible and at this point I do want to start dropping legacy java support as its becoming too complicated and we (probably more me) have been dragging our feet on moving up.

**Comment by hazendaz** [comment]
per java, this release will remain java 8.
we will though likely merge java 11 stuff right after this and if we need to patch can back port a release or two if we run into any significant issues.

**Comment by hazendaz** [comment]
We have decided since 4.9.0 is going to be java 11.  I know there are some issues with mac possibly and something with eclipse that need addressed before we cut release.  Also something with commons text we are waiting on.  If there are others that are just waiting on others to respond on open pull requests, if you are one of those that opened the pull request and want it in, please respond or it won't likely make the cut.  I want to see us get this out before end of 2024 as we haven't released in quite some time and don't expect to stay at java 11 very long provided we don't' have a lot of issues that we have introduced that cause regression concerns.

I've also reached out to maven to help with the maven spotbugs plugin as they rewrote doxia and that is required to be upgraded for maven 4 and maven 4 is nearing release candidate status now and I'm stuck with that after months of trying. The plugin is unfortunately written in groovy so its harder for java developers to grasp but if anyone can take a look as well that would be great, there is a branch called site4 that has the code where I've gotten thus far and stuck.  Beyond that, I'd like to have that aligned and released with that fix when we drop 4.9.0.

**Reply by JuditKnoll** [reply]
I agree with the mentioned mac build fail (which is connected to the eclipse issue) and the commons-text issues being necessary for 4.9.0,  and I don't see others which are a necessity, rather a nice to have. 

Leaving some links here for easy look up:
- the issue about the mac build fail: https://github.com/spotbugs/spotbugs/issues/3099
- PR updating commons-text to 1.12.0: https://github.com/spotbugs/spotbugs/pull/3202
    - I'm not sure whether [commons-text](https://github.com/apache/commons-text) 1.12.1 will be released this year, we should not count this update as blocking, but check whether it works properly with the current master.
 - 4.9.0 milestone: https://github.com/spotbugs/spotbugs/milestone/24
 - site4 branch at spotbugs-maven-plugin: https://github.com/spotbugs/spotbugs-maven-plugin/tree/site4

**Reply by JuditKnoll** [reply]
Looked into the commons-text issue a bit, see https://github.com/spotbugs/spotbugs/issues/2686#issuecomment-2528436666.

**Reply by gtoison** [reply]
I think we should release 4.9.0 sooner than later, even if there might be known issues.
There are already a lot of changes since the previous version, users affected by the known issues can remain on 4.8

**Comment by hazendaz** [comment]
OK we should be getting release candidate for commons text issue.  Let's wrap up the release with that.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by hazendaz** [reply]
commons text issue is probably addressed now.  @JuditKnoll Can you check that its working as expected now?

**Reply by JuditKnoll** [reply]
Unfortunately with older version of Spring, it still doesn't work properly, since Spring forces the usage of specific versions, which overwrites our direct and indirect dependencies.
However, with newer versions of Spring, the latest master works flawlessly. Sometime this issue should be properly fixed, but until then this is not blocking to the release IMO. This year I won't be online or be able to contribute, but do not let this stop you from releasing, if you have the time to do it.

**Reply by iloveeclipse** [reply]
Happy new year to all!
What is the current state after holidays season? Is a new release possible or do we have any blocker?

**Comment by hazendaz** [comment]
Pretty sure no blockers.  If all the github actions are showing good correctly then it's just been lack of available time to release.  I can spin it tonight as I've freed up and got time plus I'm also ready to use it now too.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by iloveeclipse** [reply]
Perfect, thanks.

**Reply by hazendaz** [reply]
There are some build issues with eclipse causing timeouts which we knew about and some warnings in the builds that I'd like to take a little more time to look at.  Its after midnight now and waiting on 2 outstanding pull requests I have out and a few that will be behind that. I don't expect to have any troubles getting released but want to wait 1 or 2 days to look at the issues and address few items that I have.  I was not that caught up but now I'm fully caught up and so are all the branches we have hanging out on the upstream here.  So lets expect Wednesday the release for 4.9.0 occurs.

**Reply by hazendaz** [reply]
If possible @iloveeclipse can you confirm the latest eclipse snapshot is working as expected?  I was hoping what I found fixed the issues with mac runner build but its still an issue.  I've found some other areas where java 8 still listed so I'll clear out all my cache tomorrow and try to catch any last bits of that but I would expect given changes that Eclipse should be compiling the plugin to 11 now.   Would be great if you can confirm that bit.

**Reply by iloveeclipse** [reply]
> If possible @iloveeclipse can you confirm the latest eclipse snapshot is working as expected?

It does. Did a smoke test for most of functionality, works as expected (build 4.9.0.202501150555-8bc2966 on latest SDK nightly build/Java 21).

**Comment by hazendaz** [comment]
@gtoison @JuditKnoll Last call for anything else to go into 4.9.0 release.  Tomorrow night is going to be the night.  I did punt a couple of things around annotations to version 5 and think we highly need to do something to rid ourselves of jsr305 at that time.  I haven't looked into anything else outstanding.  I'll poke around tomorrow evening a bit but try to confirm back if we are all good here.  As soon as the Eclipse plugin is confirmed ok, its basically a go to run the release but that may get punted as far as issue with mac runner.  This is certainly a big one in comparison to the past so I'd expect there will be discoveries we need to deal with for a week or two after unless we are just really more successful than the 4.8.x track.  For the maven / gradle plugin, I'll follow those shortly up with release but likely will be Thursday for those.

One thing we may want to additionally do is just spot check the commit changes since 4.8.6 and make sure we listed all we want in the change log.  I have everything tagged to the milestone and generally assigned to someone throughout so its easier to find everything.  A lot of the underlying general code items didn't get much for change logs which is fine but its worth checking in case we think we need to call things out.  For example, while we state java 11 now, we may want to call that out with Eclipse plugin affected as well in case someone is really using a crazy old copy ;)

**Reply by iloveeclipse** [reply]
Just a side note: Eclipse expects minimal Java 11 version since 4.21 (2021-09) release, Java 17 since 4.28 (2023-06) and it looks like we will require 21 from 4.35 (at least parts of SDK are already requiring that). 

So from my POV, there is no reason to worry about Java 11 or 17 switch in the Eclipse plugin.

**Comment by hazendaz** [comment]
I've released 4.9.0, enjoy!

**Reply by jochenberger** [reply]
When upgrading the Gradle plugin from 6.1.0 to 6.1.1, I get
```
> Task :spotbugsMain
edu.umd.cs.findbugs.ba.AnalysisException: Exception was thrown during analysis
        at edu.umd.cs.findbugs.FindBugs2.lambda$analyzeApplication$3(FindBugs2.java:1117)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at edu.umd.cs.findbugs.FindBugs2.analyzeApplication(FindBugs2.java:1110)
        at edu.umd.cs.findbugs.FindBugs2.execute(FindBugs2.java:309)
        at edu.umd.cs.findbugs.FindBugs.runMain(FindBugs.java:399)
        at edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1223)
Caused by: java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: edu/umd/cs/findbugs/annotations/SuppressMatchType
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at edu.umd.cs.findbugs.FindBugs2.lambda$analyzeApplication$3(FindBugs2.java:1112)
        ... 5 more
Caused by: java.lang.NoClassDefFoundError: edu/umd/cs/findbugs/annotations/SuppressMatchType
        at edu.umd.cs.findbugs.detect.NoteSuppressedWarnings.visitAnnotation(NoteSuppressedWarnings.java:91)
        at edu.umd.cs.findbugs.visitclass.AnnotationVisitor.visitAnnotation(AnnotationVisitor.java:315)
        at org.apache.bcel.classfile.Annotations.accept(Annotations.java:84)
        at edu.umd.cs.findbugs.visitclass.PreorderVisitor.doVisitMethod(PreorderVisitor.java:323)
        at edu.umd.cs.findbugs.visitclass.PreorderVisitor.visitJavaClass(PreorderVisitor.java:405)
        at org.apache.bcel.classfile.JavaClass.accept(JavaClass.java:249)
        at edu.umd.cs.findbugs.detect.NoteSuppressedWarnings.visitClassContext(NoteSuppressedWarnings.java:81)
        at edu.umd.cs.findbugs.DetectorToDetector2Adapter.visitClass(DetectorToDetector2Adapter.java:76)
        at edu.umd.cs.findbugs.FindBugs2.lambda$analyzeApplication$1(FindBugs2.java:1100)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at edu.umd.cs.findbugs.CurrentThreadExecutorService.execute(CurrentThreadExecutorService.java:86)
        at java.base/java.util.concurrent.AbstractExecutorService.invokeAll(AbstractExecutorService.java:247)
        ... 4 more
Caused by: java.lang.ClassNotFoundException: edu.umd.cs.findbugs.annotations.SuppressMatchType
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        ... 16 more
 ```
The class is apparently present in the annotations jar though.

**Reply by jochenberger** [reply]
When running with `-i`, I see that the old 4.8.6 annotations jar is being used:
```
Starting process 'command '/usr/lib/jvm/java-21-openjdk-amd64/bin/java''. Working directory: /home/jochen/git/test Command: /usr/lib/jvm/java-21-openjdk-amd64/bin/java -Dfile.encoding=UTF-8 -Duser.country=DE -Duser.language=de -Duser.variant -cp /home/jochen/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/2.0.0/bf7158cef919bb65a6de45766fcf0546a88f5907/slf4j-simple-2.0.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/2.0.16/172931663a09a1fa515567af5fbef00897d3c04/slf4j-api-2.0.16.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/com.github.spotbugs/spotbugs/4.9.0/f36d3c8b7e096a97ec695f7e1b77db29aae893a/spotbugs-4.9.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/jaxen/jaxen/2.0.0/bd6a33b0fda054a5678010df843cc999f288dc6c/jaxen-2.0.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/net.sf.saxon/Saxon-HE/12.5/57c007520e2879387b8d13d0a512e9566eeffa73/Saxon-HE-12.5.jar:*/home/jochen/.gradle/caches/modules-2/files-2.1/com.github.spotbugs/spotbugs-annotations/4.8.6/1dcffed3e561ed32134a0dff4717f19bc2fdf4d8/spotbugs-annotations-4.8.6.jar*:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-core/2.24.3/7f6a261243ca767c7f38fd4b542bcde626c8894e/log4j-core-2.24.3.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-commons/9.7.1/406c6a2225cfe1819f102a161e54cc16a5c24f75/asm-commons-9.7.1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-util/9.7.1/9e23359b598ec6b74b23e53110dd5c577adf2243/asm-util-9.7.1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-analysis/9.7.1/f97a3b319f0ed6a8cd944dc79060d3912a28985f/asm-analysis-9.7.1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-tree/9.7.1/3a53139787663b139de76b627fca0084ab60d32c/asm-tree-9.7.1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm/9.7.1/f0ed132a49244b042cd0e15702ab9f2ce3cc8436/asm-9.7.1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.bcel/bcel/6.10.0/34f0f3f24d2cd6627e304cbdd5ce2b9ca0671e4c/bcel-6.10.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/com.github.stephenc.jcip/jcip-annotations/1.0-1/ef31541dd28ae2cefdd17c7ebf352d93e9058c63/jcip-annotations-1.0-1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.dom4j/dom4j/2.1.4/35c16721b88cf17b8279fcb134c0abb161cc0e9b/dom4j-2.1.4.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-text/1.13.0/ba2ed5521c491cabf7ecdb57f77922561c2e8958/commons-text-1.13.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-lang3/3.17.0/b17d2136f0460dcc0d2016ceefca8723bdf4ee70/commons-lang3-3.17.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.xmlresolver/xmlresolver/5.2.2/152378e04ba01898847bf38fad5023c7d52f0c9d/xmlresolver-5.2.2.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.xmlresolver/xmlresolver/5.2.2/5624ede8b8e374979194acaae9f34cff23b62b3e/xmlresolver-5.2.2-data.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents.client5/httpclient5/5.4.1/ce913081e592ee8eeee35c4e577d7dce13cba7a4/httpclient5-5.4.1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/com.google.code.gson/gson/2.11.0/527175ca6d81050b53bdd4c457a6d6e017626b0e/gson-2.11.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.logging.log4j/log4j-api/2.24.3/b02c125db8b6d295adf72ae6e71af5d83bce2370/log4j-api-2.24.3.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/com.google.code.findbugs/jsr305/3.0.2/25ea2e8b0c338a877313bd4672d3fe056ea78f0d/jsr305-3.0.2.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/com.google.errorprone/error_prone_annotations/2.27.0/91b2c29d8a6148b5e2e4930f070d4840e2e48e34/error_prone_annotations-2.27.0.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents.core5/httpcore5-h2/5.3.1/760c34db3ba41b0ffa07e956bc308d3a12356915/httpcore5-h2-5.3.1.jar:/home/jochen/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents.core5/httpcore5/5.3.1/eaf64237945d7d0f301d48420e8bdb7f565a7b0e/httpcore5-5.3.1.jar edu.umd.cs.findbugs.FindBugs2 -exitcode -pluginList /home/jochen/.gradle/caches/modules-2/files-2.1/com.h3xstream.findsecbugs/findsecbugs-plugin/1.13.0/f6c3030af89489e6d1a745e2accc139dba459e52/findsecbugs-plugin-1.13.0.jar -timestampNow -auxclasspathFromFile /home/jochen/git/test/build/spotbugs/auxclasspath/spotbugsMain -sourcepath /home/jochen/git/test/src/main/resources:/home/jochen/git/test/src/main/java:/home/jochen/git/test/src/main/groovy -html=/home/jochen/git/test/build/reports/spotbugs/main.html -exclude /home/jochen/git/test/misc/spotbugs/exclude.xml -onlyAnalyze  -projectName test (spotbugsMain) -release 1.0 -analyzeFromFile /home/jochen/git/test/build/spotbugsMain-analyse-class-file.txt
```

**Reply by gtoison** [reply]
Do you declare the spotbugs dependency somewhere (on top of declaring the spotbugs gradle plugin)?

**Reply by jochenberger** [reply]
I don't think so. I'm using Spring Boot though, so you never know. :face_with_diagonal_mouth:

**Reply by jochenberger** [reply]
Yeah, it seems to be Spring Boot. I suspect that it's this issue: https://github.com/spring-projects/spring-boot/issues/42522, or rather that one: https://github.com/apache/logging-log4j2/issues/3066

**Reply by gtoison** [reply]
That makes sense yes, so you probably need to force the spotbugs dependency to use version 4.9.0

**Comment by gtoison** [comment]
What about releasing 4.9.1 now with the fixes for #3266 and #3267 ?

**Reply by iloveeclipse** [reply]
No objections. So far we only had those two regressions reported & fixed, so it would be a nice point release.

**Reply by iloveeclipse** [reply]
With https://github.com/spotbugs/spotbugs/issues/3280 fix it would be nice to have 4.9.1 soon.

**Reply by hazendaz** [reply]
We still good to move on 4.9.1?  I can do it Friday night.

**Reply by iloveeclipse** [reply]
I believe all known regressions are fixed. I've created milestone to see what is relevant: https://github.com/spotbugs/spotbugs/milestone/25?closed=1

**Reply by gtoison** [reply]
The release workflow apparently failed for 4.9.1

**Reply by iloveeclipse** [reply]
Correct. What was the problem? Should we now try to release 4.9.2?

**Reply by JuditKnoll** [reply]
It looks like to the `Commit CHANGELOG.md` ([in the workflow file](https://github.com/spotbugs/spotbugs/blob/master/.github/workflows/release.yml#L38)) failed (see the [failing job](https://github.com/spotbugs/spotbugs/actions/runs/13105274127/job/36558993588)). It was skipped earlier (see [4.9.0 release job](https://github.com/spotbugs/spotbugs/actions/runs/12802354965/job/35693306634)). It's at quite the beginning of the release workflow, and all the later steps was skipped, so nothing really happened. I think if the 4.9.1 tag is deleted, then pushed back, maybe it will work.

**Reply by iloveeclipse** [reply]
> I think if the 4.9.1 tag is deleted, then pushed back, maybe it will work.

Could you ... ?

**Reply by JuditKnoll** [reply]
> > I think if the 4.9.1 tag is deleted, then pushed back, maybe it will work.
> 
> Could you ... ?

Done, added the changelog text and published the 4.9.1 release on GitHub.

**Reply by gtoison** [reply]
4.9.1 does not seem to be on maven central unfortunately, I don't see it here: https://repo1.maven.org/maven2/com/github/spotbugs/spotbugs/

**Reply by JuditKnoll** [reply]
It looks like the `Build otherwise` step run instead of the `Build on tag`, which doesn't run `closeAndReleaseSonatypeStagingRepository`, but it should publish it to SonaType as both `publishToSonatype` and `createReleaseBody` run. This part is the same as it was for the release of 4.9.1. I can't find 4.9.1 in [SonaType](https://oss.sonatype.org/), only 4.9.1-SNAPSHOT, but that was uploaded on 2nd/3rd of Februrary, and it's a tgz, while the other artifacts are zips. The Staging repositories page is also empty on Sonatype.
In the [log of the release job](https://github.com/spotbugs/spotbugs/actions/runs/13162401876/job/36734198617) I couldn't find any indication of Sonatype issues. Can someone please look into it?

**Reply by hazendaz** [reply]
I tried to run the tag again to see what would happen, it again just went the else path.  Not sure what the issue is.

**Reply by hazendaz** [reply]
I tried to rerun in debug, it suddenly worked.  There must be some issue with sonatype.  Looking at the GHA, its all wrong as it doesn't work as designed (areas never ran), probably needs some cleanup, but looking at 4.9.0 vs 4.9.1, they both ran the exact same things.  4.9.0 worked whereas 4.9.1 on multiple attempts did not but then today it did.  Nothing else was done. Its released now though.  It should show in central shortly.

**Comment by hazendaz** [comment]
It's staged in sonatype.  Needs to be releases there.  I'll get it today

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by garydgregory** [comment]
Maybe upload the JAR manually using the Nexus UI until you sort this out for the next release?

**Comment by garydgregory** [comment]
I can see it now in Maven Central, TY all.
